"""
==============================
Snakemake annotation pipeline
==============================
- USAGE - locally - whilst using the annotation environment
    snakemake --configfile config_file.yaml --cores 10
- USAGE - cluster - whilst using the annotation environment
    snakemake --configfile config_file.yaml --cores 10 --cluster-config
    ./cluster.yaml --cluster "bsub -q {cluster.queue} -oo {cluster.output}
    -eo {cluster.error} -M {cluster.memory} -R {cluster.resources} -J {cluster.jobname}"
    -j 10 --use-conda

shamelessly stolen from Damon
"""

import os
import errno

reference = config['genome']
analysis  = config['analysis']
mydir = config['working_dir']
analysis_dict = {}

for key, value in analysis.items():
    directory = os.path.join(mydir, key)
    for param, vvlaue in value.items():
        analysis_dict[key] = directory
        try:
            print(f'Working dir: {directory}')
            os.makedirs(directory)
        except OSError as e:
            if e.errno != errno.EEXIST:
                raise

rule all:
   input: 's2_braker/s2_done' # needs changing to s4 later

rule s0_bam2fastq:
    input:
      analysis['s0_bam2fastq']['bam']
    output:
      output: a = "s0_bam2fastq/R1.fq", b = "s0_bam2fastq/R2.fq", c = "s0_bam2fastq/R0.fq" # not sure if we should cater the non-paired case
    shell:
        "samtools fastq -1 {output.a} -2 {output.b} -0 {output.c} {input}"

rule s1_alignment:
    input:
      reference,
      "s0_bam2fastq/R1.fq", 
      "s0_bam2fastq/R2.fq",
      analysis['s1_alignment']['output']
    shell:
      "bash ../braker/star.bash {input} {output[0]}" #need to change the script to take optionally paired reads 

rule s2_braker:
     input:
       reference,
       rules.s1_alignment.output,
     output:
       "s2_braker/s2_done"
     params:
        species = config['species']
     shell:
       "bash ../braker/braker_test.bash {input} {params.species};"
       " touch s2_braker/s2_done"
# s3_make_stats

# s4_make_hub